Use decision trees for code generation on the js target

giacomocavalieri opened the pull request:
This PR closes #3969, closes #4221 and closes #4387

This is still a WIP as I haven't taken care of bit array patterns yet, but it's ready for everything else if anyone fancies giving this a look in the meantime.

What's missing:
- [X] Use the decision tree for cases
- [X] Reuse the decision tree for let restructuring and let assert 
- [x] Support bit arrays in the decision tree

giacomocavalieri commented on compiler-core/src/javascript/expression.rs:
```diff
@@ -1411,6 +1302,490 @@ impl<'module, 'a> Generator<'module, 'a> {
+        Ok(SizedBitArraySegmentDetails { size, size_value })
+    }
+
+    pub(crate) fn guard(&mut self, guard: &'a TypedClauseGuard) -> Output<'a> {
```
All the code here up to line ~1900 is basically moved as-is from the pattern module where it was previously

lpil commented:
Please give PRs titles, just a number means folks have to look them up to know what this is

giacomocavalieri commented:
Oops my mistake!

giacomocavalieri commented:
This is ready for review now!

lpil commented on compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__strings__string_prefix_assignment_with_multiple_subjects.snap:
```diff
@@ -17,9 +17,6 @@ export function go(x) {
     return prefix;
-  } else if (x.startsWith("11")) {
-    let prefix = "11";
-    return prefix;
```
Wow

giacomocavalieri replied:
Mind blowing, innit?

lpil commented on compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__case__case_on_error.snap:
```diff
@@ -0,0 +1,30 @@
+
+export function main() {
+  let $ = a_result();
+  if (!$.isOk()) {
```
I think we can remove the special case for result and use instanceof like the other variants?

giacomocavalieri replied:
We would still need to special case results as we'd have to make sure that we import the constructors from the prelude to perform the `instanceOf` check. But if you want I can switch to instance of checks for these as well

lpil replied:
Yes please

giacomocavalieri commented on compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__case_clause_guards__keyword_var.snap:
```diff
@@ -34,18 +34,26 @@ export function main() {
   } else {
-    let class$1 = var$;
-    return false;
+    if (isEqual(toList([class$]), toList([5]))) {
```
Not in general, sometimes there could be variable definitions between that `else` and the next `if`, making it impossible to turn it into an `else if`. It would be quite a bit of work and I don't think it would really be worth the effort

lpil replied:
I meant for the case in which there are no assignments specifically. Is that challenging?

giacomocavalieri replied:
I've opened an issue for this! https://github.com/gleam-lang/gleam/issues/4522

lpil commented on compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__match_literal_unaligned_float.snap:
```diff
@@ -0,0 +1,33 @@
+  let n = 1;
+  if (
+    x.bitSize < n ||
+    x.bitSize < 64 + n ||
```
Previously Gears made it not output these redundant checks, it would be fab to retain that!

lpil reviewed (commented):
Some comments as I'm half way through reviewing!

lpil commented on compiler-core/src/exhaustiveness.rs:
```diff
@@ -588,6 +589,17 @@ impl RuntimeCheck {
+    pub(crate) fn referenced_segment_patterns(&self) -> Vec<(&EcoString, &ReadAction)> {
+        match self {
+            RuntimeCheck::BitArray { test } => test.referenced_segment_patterns(),
+            _ => vec![],
```
No catch alls please

lpil commented on compiler-core/src/javascript/expression.rs:
```diff
@@ -120,6 +117,11 @@ pub(crate) struct Generator<'module, 'ast> {
+    /// This will be true if we've generated a statement that is guaranteed to
+    /// throw. This means we can stop code generation for all following statements
+    /// in the same block!
+    pub statement_always_throws: bool,
```
I think this should be set to true for `panic` and `todo`. I can't see that having been done, has it?

lpil commented on compiler-core/src/ast.rs:
```diff
@@ -3028,6 +3093,10 @@ pub struct Assignment<TypeT, ExpressionT> {
+    pub compiled_case: CompiledCase,
+    /// This will be true for assignments that are automatically generated by
+    /// the compiler.
+    pub is_generated: bool,
```
How come this is a field now?

giacomocavalieri commented on compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__case_match_literal_float.snap:
```diff
@@ -0,0 +1,33 @@
+function go(x) {
+  if (x.bitSize >= 64) {
+    if (bitArraySliceToFloat(x, 0, 64, true) === 1.4) {
+      if (x.bitSize === 72) {
```
I've opened an issue for this one! https://github.com/gleam-lang/gleam/issues/4524

giacomocavalieri commented:
I've rebased on main adding support for aliased segments for the bit array code generation. I've also found a small bug in the analysis of bit array aliased segments where it would wrongly accept a variable sized utf8 string. I've fixed it making sure it's rejected!

Norlock commented:
Thanks a lot for all the work, how much will this roughly improve on js?

giacomocavalieri commented:
Since this is improving the pattern matching logic it will depend on how much your code relies on complex pattern matching expressions. We've tested this on Lustre's DOM diffing algorithm (which is now written entirely in Gleam) and measured an impressive ~30% speedup!

I've also been benchmarking this on the todomvc lustre app and could measure a ~20% speedup, to me it's amazing seeing this kind of speedup on a "real world" app and not just in benchmarks!

Norlock commented:
Ok that's insane! I have a big Lustre app, so knowing this is really nice. After Lustre v5 and updates here my app will feel really fast.
